roboteam_ai/src/utilities/OffensiveCoach.cpp

```diff
@@ -0,0 +1,312 @@
+
+double OffensiveCoach::calculateCloseToGoalScore(Vector2 position) {
+    double distanceFromGoal = (Field::get_their_goal_center() - position).length();
+
+    double score = exp(-0.1 * distanceFromGoal);
+    return score;
+}
+
+double OffensiveCoach::calculateShotAtGoalScore(Vector2 position, roboteam_msgs::World world) {
+    double safelyness = 3;
+    while (safelyness > 0) {
+            if (control::ControlUtils::clearLine(position, Field::get_their_goal_center(), world, safelyness)) {
+                break;
+            }
+            safelyness -= 0.5;
+        }
+
+    return 1 - exp(-safelyness);
+}
+
+
+double OffensiveCoach::calculatePassLineScore(Vector2 position, roboteam_msgs::World world) {
+    double safelyness = 3;
+    while (safelyness > 0) {
+        if (control::ControlUtils::clearLine(world.ball.pos, position, world, safelyness)) {
+            break;
+        }
+        safelyness -= 0.5;
+    }
+
+    return 1 - exp(-safelyness);
+}
+
+double OffensiveCoach::calculateDistanceToOpponentsScore(Vector2 position, roboteam_msgs::World world) {
+    shared_ptr<roboteam_msgs::WorldRobot> closestRobot = World::getRobotClosestToPoint(world.them, position);
+    if (closestRobot) {
+        double distance = (position - closestRobot->pos).length();
+        return 1 - exp(-0.5 * distance);
+    } else {
+        return 1;
+    }
+}
+
+double OffensiveCoach::calculateDistanceFromCorner(Vector2 position, roboteam_msgs::GeometryFieldSize field) {
+    Vector2 corner;
+    corner.x = field.field_length / 2;
+    if (position.y > 0) {
+        corner.y = field.field_width / 2;
+    } else {
+        corner.y = -field.field_width / 2;
+    }
+    double distanceFromCorner = (position - corner).length();
+    return 1 - exp(-0.05 * distanceFromCorner);
+}
+
+double OffensiveCoach::calculateDistanceFromBallScore(Vector2 position, roboteam_msgs::GeometryFieldSize& field, roboteam_msgs::WorldBall& ball) {
+    double distanceFromBall = (position - ball.pos).length();
+    return -pow(distanceFromBall / (0.5 * maxDistanceFromBall), 2) + 2 * (distanceFromBall / (0.5 * maxDistanceFromBall));
+}
+
+double OffensiveCoach::calculatePositionScore(Vector2 position) {
+    roboteam_msgs::World world = World::get_world();
+    roboteam_msgs::GeometryFieldSize field = Field::get_field();
+    double closeToGoalScore = calculateCloseToGoalScore(position);
+    double shotAtGoalScore = calculateShotAtGoalScore(position, world);
+    double passLineScore = calculatePassLineScore(position, world);
+    double closestOpponentScore = calculateDistanceToOpponentsScore(position, world);
+    double distanceFromBallScore = calculateDistanceFromBallScore(position, field, world.ball);
+    double behindBallScore = position.x < world.ball.pos.x ? 0.7 : 1.0;
+    double distanceFromCornerScore = calculateDistanceFromCorner(position, field);
+
+    double score = 2 * closeToGoalScore + 2 * shotAtGoalScore + passLineScore + closestOpponentScore
```

Comment:
magic

Reply:
Well yeah in that sense the whole scoring is magic